### PR TITLE
Web Module

### PR DESCRIPTION
**Pull request by dmdashenkov**
## Spine `web` API

Introducing Spine `web` API - the HTTP API for a Spine-based application.
The library defines two endpoints:
 - __command__ endpoint, defined by the `CommandServlet`;
 - __query__ endpoint, defined by the `QueryServlet`.

The servlets dispatch messages into the `Command`- or `QueryService` respectively.

The `CommandServlet` always responds with an `Ack` message, encoded as a [JSON](https://developers.google.com/protocol-buffers/docs/proto3#json).

The `QueryServlet` may process the queries asynchronously. The format of the response is implementation-specific.

## `firebase-web` library

`firebase-web` library is an implementation of the Spine `web` API based on the [Fireabse Realtime Database](https://firebase.google.com/docs/database/).

The database acts as the channel for delivering query results to a client. See the [README](https://github.com/SpineEventEngine/gcloud-java/blob/web/firebase-web/README.md) for more info.

---

### Java Version

Starting from this PR, the repository uses Java 8 instead of Java 7.

The new tests (the tests for `web` and `firebase-web` modules) are written using JUnit 5. The old tests (the tests for the `datastore` module) are still written using JUnit 4. The JUnit Vintage engine is used to launch those tests.

### Repository Name

As this repository is renamed from `gae-java` to `gcloud-java`, the settings are changed respectively.

### Spine Version

The repository is migrated to Spine of version `0.10.35-SNAPSHOT`.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: web/src/main/java/io/spine/web/package-info.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright (c) 2000-2018 TeamDev Ltd. All rights reserved.
+ * TeamDev PROPRIETARY and CONFIDENTIAL.
+ * Use is subject to license terms.
+ */
+
+/**
+ * This package contains the basic client transport API.
```
Comment (armiol) [inline web/src/main/java/io/spine/web/package-info.java]: I don't think this package is only about a transport.

Also, please rephrase `basic`.

File: web/src/main/java/io/spine/web/QueryMediator.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * <p>Translates the given {@link Query} into a {@link QueryResult}.
+ *
+ * <p>No constrains are applied to the contents of the query. Neither any guaranties are made for
+ * the query result. Investigate the concrete implementation to find out the details of their
+ * behaviour.
+ *
+ * @author Dmytro Dashenkov
+ */
+public interface QueryMediator {
```
Comment (armiol) [inline web/src/main/java/io/spine/web/QueryMediator.java]: Let's rename it to `QueryBridge` as discussed. And tweak the documentation accordingly.

File: web/src/main/java/io/spine/web/parser/package-info.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+/*
+ * Copyright (c) 2000-2018 TeamDev Ltd. All rights reserved.
+ * TeamDev PROPRIETARY and CONFIDENTIAL.
+ * Use is subject to license terms.
+ */
+
+/**
+ * This package contains components responsible for parsing messages sent over the Web protocols.
```
Comment (armiol) [inline web/src/main/java/io/spine/web/parser/package-info.java]: Not sure what you mean by `Web protocols`. And why `Web` is capitalized.

**Review by armiol — COMMENTED** [review]
@dmdashenkov please see my comments.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: build.gradle [hunk 4]
```diff
@@ -93,70 +104,73 @@ subprojects {
...
-
-codeCoverageReport.dependsOn {
-    subprojects*.test
-}
-
 task startDatastore {
-    description "Starts local in-memory datastore using scripts form ./scripts folder."
-    group "Build Setup"
-}
+    description = "Starts local in-memory datastore using scripts form ./scripts folder."
```
Comment (armiol) [inline build.gradle]: `from`.

File: build.gradle [hunk 5]
```diff
@@ -93,70 +104,73 @@ subprojects {
...
+    group = "Build Setup"
 
-startDatastore.doLast {
-    // Start Datastore emulator with the CLI command
-    ".${File.separatorChar}scripts${File.separatorChar}start-datastore.${runsOnWindows ? 'bat' : 'sh'}".execute()
+    doLast {
+        // Start Datastore emulator with the CLI command
```
Comment (armiol) [inline build.gradle]: Please append this line with a period.

File: firebase-web/README.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+```
+
+This JSON is parsed back into the `Query` message by the `QueryServlet` and sent to 
+the `QueryService` for an asynchronous processing.
+
+Alternatively, the `Query` may be represented as a Base64-encoded byte string. 
+See [`HttpMessages`](../client-transport/src/main/java/com/teamdev/licensing/web/parser/HttpMessages.java)
+for more info. 
+
+The response path is generated based on the `Query`. Assume that the path is `foo/bar`.
```
Comment (armiol) [inline firebase-web/README.md]: `basing` would look better. Or `The response path is based on ...`.

File: firebase-web/build.gradle [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+ext {
+    firebaseApiVersion = '5.9.0'
```
Comment (armiol) [inline firebase-web/build.gradle]: Please extract to `ext.gradle`.

File: firebase-web/src/main/java/io/spine/web/firebase/FirebaseQueryBridge.java [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+         *
+         * @return new instance of {@code FirebaseQueryBridge}
+         */
+        public FirebaseQueryBridge build() {
+            checkState(queryService != null, "Query Service is not set.");
+            checkState(database != null, "FirebaseDatabase is not set.");
+            return new FirebaseQueryBridge(this);
+        }
+    }
+
```
Comment (armiol) [inline firebase-web/src/main/java/io/spine/web/firebase/FirebaseQueryBridge.java]: This line is redundant.

File: firebase-web/src/main/java/io/spine/web/firebase/FirebaseQueryBridge.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+import io.spine.client.grpc.QueryServiceGrpc.QueryServiceBlockingStub;
+import io.spine.client.grpc.QueryServiceGrpc.QueryServiceImplBase;
+
+import java.util.concurrent.CompletableFuture;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * A Firebase Realtime Database based implementation of {@link QueryBridge}.
```
Comment (armiol) [inline firebase-web/src/main/java/io/spine/web/firebase/FirebaseQueryBridge.java]: `An implementation of ... based on ...`

File: firebase-web/src/main/java/io/spine/web/firebase/FirebaseQueryBridge.java [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+ *
+ * <p>This bridge stores the {@link QueryResponse} data to a location in a given
+ * {@link FirebaseDatabase} and retrieves the database path to that response as the result.
+ *
+ * <p>More formally, for each encountered {@link Query}, the bridge performs a call to
+ * the {@code QueryService} and stores the resulting entity states into the given database. The data
+ * is stored as a list of strings. Each entry is
+ * a {@linkplain io.spine.json.Json JSON representation} of an entity state. The path produced by
+ * the bridge as a result is the path to the database node containing all those records.
+ * The absolute position of such node is not specified, thus the result path is the only way to read
```
Comment (armiol) [inline firebase-web/src/main/java/io/spine/web/firebase/FirebaseQueryBridge.java]: `such a node`

File: firebase-web/src/main/java/io/spine/web/firebase/FirebaseQueryBridge.java [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+    private final long writeAwaitSeconds;
+
+    private FirebaseQueryBridge(Builder builder) {
+        this.queryService = builder.queryService;
+        this.database = builder.database;
+        this.writeAwaitSeconds = builder.writeAwaitSeconds;
+    }
+
+    @Override
+    public QueryProcessingResult send(Query query) {
```
Comment (armiol) [inline firebase-web/src/main/java/io/spine/web/firebase/FirebaseQueryBridge.java]: Please either use `{@inheritDoc}` or, as I suspect, add some details.

File: firebase-web/src/main/java/io/spine/web/firebase/package-info.java [hunk 12]
```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright (c) 2000-2018 TeamDev Ltd. All rights reserved.
+ * TeamDev PROPRIETARY and CONFIDENTIAL.
+ * Use is subject to license terms.
+ */
+
+/**
+ * This package contains the Firebase client transport.
```
Comment (armiol) [inline firebase-web/src/main/java/io/spine/web/firebase/package-info.java]: The `transport` itself? Really?

File: web/src/main/java/io/spine/web/FutureObserver.java [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+import io.grpc.stub.StreamObserver;
+
+import javax.annotation.Nullable;
+import java.util.concurrent.CompletableFuture;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.spine.util.Exceptions.newIllegalStateException;
+
+/**
+ * An unary {@linkplain StreamObserver stream observer} which can provide
```
Comment (armiol) [inline web/src/main/java/io/spine/web/FutureObserver.java]: I wouldn't call it `unary` as all of them are.

**Review by armiol — COMMENTED** [review]
@dmdashenkov please see my comments.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

**Review by armiol — APPROVED** [review]
@dmdashenkov LGTM.

**Pull request merged**